tests/stdlib/tasynchttpserver_transferencoding.nim

```diff
@@ -0,0 +1,150 @@
+import httpclient, asynchttpserver, asyncdispatch, asyncfutures
+import net
+
+const postBegin = """
+POST / HTTP/1.1
+Host: 127.0.0.1:64123
+Accept: */*
+Transfer-Encoding:chunked
+Content-Type: application/x-www-form-urlencoded
+
+"""
+var
+    sanitySimpleHelloWorldTest = false
+    sanitySimpleEncodingTest = false
+    sanitySimpleChunksTest = false
+    sanityComplexChunksTest = false
+
+proc testSimpleHelloWorld() =
+    let exampleWorld = ("b\r\n" &
+                        "hello=world\r\n" &
+                        "0\r\n" &
+                        "\r\n")
+    let expected = "hello=world"
+
+    proc handler(request: Request) {.async.} =
+        doAssert(request.body == expected)
+        doAssert(request.headers.hasKey("Transfer-Encoding"))
+        doAssert(not request.headers.hasKey("Content-Length"))
+        sanitySimpleHelloWorldTest = true
+        await request.respond(Http200, "Good")
+
+    let server = newAsyncHttpServer()
+    discard server.serve(Port(64123), handler)
+
+    let data = postBegin & exampleWorld
+
+    var socket = newSocket()
+    socket.connect("127.0.0.1", Port(64123))
+    socket.send(data)
+
+    waitFor sleepAsync(10)
+
+    socket.close()
+    server.close()
+
+proc testSimpleEncoding() =
```

Comment:
these tests are basically copy paste of some common template. This is bad as it makes evolving such tests harder that it should. Instead, use templates for that, with params corresponding to whatever changes between each test. eg:

```nim
template genTest(input, expected) =
  proc handler(request: Request) {.async.} =
      doAssert(request.body == expected)
      doAssert(request.headers.hasKey("Transfer-Encoding"))
      doAssert(not request.headers.hasKey("Content-Length"))
      sanityComplexChunksTest = true
      await request.respond(Http200, "Good")

  let server = newAsyncHttpServer()
  discard server.serve(Port(64123), handler)

  let data = postBegin & input

  var socket = newSocket()
  socket.connect("127.0.0.1", Port(64123))
  socket.send(data)

  waitFor sleepAsync(10)

  socket.close()
  server.close()

block:
  # https://en.wikipedia.org/wiki/Chunked_transfer_encoding#Example
  let expected = "Wikipedia in \r\n\r\nchunks."
  let exampleWikipedia = ("4\r\n" &
                          "Wiki\r\n" &
                          "6\r\n" &
                          "pedia \r\n" &
                          "E\r\n" &
                          "in \r\n" &
                          "\r\n" &
                          "chunks.\r\n" &
                          "0\r\n" &
                          "\r\n")
  genTest(input, expected)
block:
  # ... etc
  genTest(input, expected)
```

Reply:
Beautiful! Fixed.